astropy/coordinates/name_resolve.py

```diff
@@ -152,12 +167,58 @@ def get_icrs_coordinates(name, parse=False):
             fmt_url = fmt_url.format(name=urllib.parse.quote(name), db=db)
             urls.append(fmt_url)
 
+    if cache:
+        try:
+            dldir, urlmapfn = _get_download_cache_locs()
+        except OSError as e:
+            msg = 'Remote data cache could not be accessed due to '
+            estr = '' if len(e.args) < 1 else (': ' + str(e))
+            warn(data.CacheMissingWarning(msg + e.__class__.__name__ + estr))
+            cache = False
+
     exceptions = []
     for url in urls:
+
+        if cache and not overwrite:
+            with shelve.open(urlmapfn) as url2hash:
+                if url in url2hash:
+                    with open(url2hash[url], 'rb') as f:
+                        resp_data = f.read()
+                        break
+
         try:
             # Retrieve ascii name resolve data from CDS
             resp = urllib.request.urlopen(url, timeout=data.conf.remote_timeout)
             resp_data = resp.read()
+
+            if cache:
+                # Keep a hash to rename the local file to the hashed name.
+                # Here, other download functions will hash on the response data,
+                # but simbad/sesame intermittently add and remove some
+                # extraneous content from the response (e.g., the string "from
+                # cache" and some other table information), so here we instead
+                # hash the URL:
+                hash = hashlib.md5()
+                hash.update(url.encode('utf-8'))
+
+                _acquire_download_cache_lock()
+
+                try:
+                    with shelve.open(urlmapfn) as url2hash:
+                        # We check now to see if another process has
+                        # inadvertently written the file underneath us
+                        if url in url2hash and not overwrite:
+                            with open(url2hash[url], 'rb') as f:
+                                resp_data = f.read()
+                                break
+
+                        local_path = os.path.join(dldir, hash.hexdigest())
+                        with open(local_path, 'wb') as f:
+                            f.write(resp_data)
```

Comment:
This entire loop body could be replaced by `open(astropy.utils.download_file(url, cache=cache, overwrite=overwrite), "rb").read()` if download_file supported `overwrite`. That would be a fairly easy addition, particularly once #9182 is merged.